core/block/editor/profilemigration/profilemigration.go

```diff
@@ -0,0 +1,97 @@
+package profilemigration
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/anyproto/anytype-heart/core/block/editor/state"
+	"github.com/anyproto/anytype-heart/core/block/simple"
+	"github.com/anyproto/anytype-heart/core/block/simple/text"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/pkg/lib/bundle"
+	smartblock2 "github.com/anyproto/anytype-heart/pkg/lib/core/smartblock"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+	"github.com/anyproto/anytype-heart/util/pbtypes"
+)
+
+const InternalKeyOldProfileData = "oldprofile"
+
+var ErrNoCustomStateFound = fmt.Errorf("no custom state found")
+
+// ExtractCustomState extract user-added state to the separate state and removes all the custom blocks/details from the original one
+func ExtractCustomState(st *state.State) (userState *state.State, err error) {
+	identityBlockId := "identity"
+	// we leave identity and other blocks in the original object to avoid them being re-added by old clients
+	whitelistBlocks := []string{state.HeaderLayoutID, state.DescriptionBlockID, state.FeaturedRelationsID, state.TitleBlockID, identityBlockId}
+	hasCustomState := false
+	st.Iterate(func(b simple.Block) (isContinue bool) {
+		if slices.Contains(whitelistBlocks, b.Model().Id) {
+			return true
+		}
+		if textBlock, ok := b.(text.Block); ok {
```

Comment:
maybe use IsEmpty for text block too? You can have empty text and colored background, for example

Reply:
main point was to avoid migration if there is no any valuable information.
IsEmpty() method has a lot of checks inside, including align != default. If the text is empty it doesn't make any sense.

Reply:
I could have background color for example, is it valuable information? It could be used for decorations. And layout blocks could be valuable as well. Maybe just copy everything?